### ISSUE TITLE
Question about Continual Pre-Training

### ISSUE THREAD

**Issue by hrlics** [TsinghuaC3I/Fourier-Position-Embedding#13]
Hi @Hua-rookie, thanks for sharing your implementation!

When I ran `commands/run_finetune.sh` to conduct **Continual Pre-Training** using the last checkpoint from pre-training, the process terminated immediately without actually starting.

I suspect the issue is caused by the **completed** status in the trainer state saved in the last pre-training checkpoint. As a result, when loading this checkpoint, the script reaches [this line](https://github.com/TsinghuaC3I/Fourier-Position-Embedding/blob/54faf9567668848616e30f89ee42180066dba5f5/scripts/train.py#L286) and exits early.

Setting `reset_trainer_state: true` resolved the problem on my end.

**Comment by hrlics** [comment]
Also, it seems that `commands/run_finetune.sh` uses the same amount of tokens (10B) as in pre-training, while it is reported as 1B in the paper.

**Comment by Hua-rookie** [comment]
Hi @hrlics, I will answer point by point:
 
> Hi [@Hua-rookie](https://github.com/Hua-rookie), thanks for sharing your implementation!
> 
> When I ran `commands/run_finetune.sh` to conduct **Continual Pre-Training** using the last checkpoint from pre-training, the process terminated immediately without actually starting.
> 
> I suspect the issue is caused by the **completed** status in the trainer state saved in the last pre-training checkpoint. As a result, when loading this checkpoint, the script reaches [this line](https://github.com/TsinghuaC3I/Fourier-Position-Embedding/blob/54faf9567668848616e30f89ee42180066dba5f5/scripts/train.py#L286) and exits early.
> 
> Setting `reset_trainer_state: true` resolved the problem on my end.

We did not encounter this issue in our experiments. This is because C4 is significantly larger than 10B and the training process is far from completion. I suspect that the issue may be attributed to the use of Gutenberg Books as the training set, which has a relatively limited number of tokens.

> Also, it seems that `commands/run_finetune.sh` uses the same amount of tokens (10B) as in pre-training, while it is reported as 1B in the paper.

Similarly, I suspect that you might have used Gutenberg Books instead of C4. It would be advisable to verify the dataset that was employed for training🌹

**Comment by hrlics** [comment]
Hi @Hua-rookie, thanks for your response! 😄 

I've double checked and ensured that the dataset used for pre-training was C4. I think there might be some misunderstandings. I will detail my steps as follows.

> We did not encounter this issue in our experiments. This is because C4 is significantly larger than 10B and the training process is far from completion. I suspect that the issue may be attributed to the use of Gutenberg Books as the training set, which has a relatively limited number of tokens.

1. I completed `commands/run_pretrain.sh` and saved the last checkpoint at `workspace/OLMo-60M-ce-512-c4/latest-unsharded`.
2. I ran `commands/run_finetune.sh` using the last checkpoint from pretraining:
```
torchrun --nproc_per_node=4 scripts/train.py \
         configs/c4/length-1024/ce-extra/x+yarn/OLMo-60M-ce-yarn.yaml \
         --load_path=workspace/OLMo-60M-ce-512-c4/latest-unsharded
```
3. The training completes in less than 30 seconds.

---

Regarding my Q2,
> Similarly, I suspect that you might have used Gutenberg Books instead of C4. It would be advisable to verify the dataset that was employed for training🌹

In the paper, it is stated that the **continue pretraining** uses ~1B tokens, which is less than the number of tokens used for pre-training (10B)

![Image](https://github.com/user-attachments/assets/acb0b40d-9b31-4597-ac28-7e43f70ad16b)

However, it is shown in the script that the **continue pretraining** uses the [same training data (from part-000-00000 to part-170-00000)](
https://github.com/TsinghuaC3I/Fourier-Position-Embedding/blob/main/configs/c4/length-1024/ce-extra/x%2Byarn/OLMo-60M-ce-yarn.yaml#L283
) as [**pre-training**](https://github.com/TsinghuaC3I/Fourier-Position-Embedding/blob/main/configs/c4/length-512/ce-rope/OLMo-60M-ce.yaml#L265)

**Comment by Hua-rookie** [comment]
> Hi [@Hua-rookie](https://github.com/Hua-rookie), thanks for your response! 😄
> 
> I've double checked and ensured that the dataset used for pre-training was C4. I think there might be some misunderstandings. I will detail my steps as follows.
> 
> > We did not encounter this issue in our experiments. This is because C4 is significantly larger than 10B and the training process is far from completion. I suspect that the issue may be attributed to the use of Gutenberg Books as the training set, which has a relatively limited number of tokens.
> 
> 1. I completed `commands/run_pretrain.sh` and saved the last checkpoint at `workspace/OLMo-60M-ce-512-c4/latest-unsharded`.
> 2. I ran `commands/run_finetune.sh` using the last checkpoint from pretraining:
> 
> ```
> torchrun --nproc_per_node=4 scripts/train.py \
>          configs/c4/length-1024/ce-extra/x+yarn/OLMo-60M-ce-yarn.yaml \
>          --load_path=workspace/OLMo-60M-ce-512-c4/latest-unsharded
> ```
> 
>  3. The training completes in less than 30 seconds.
> 
> Regarding my Q2,
> 
> > Similarly, I suspect that you might have used Gutenberg Books instead of C4. It would be advisable to verify the dataset that was employed for training🌹
> 
> In the paper, it is stated that the **continue pretraining** uses ~1B tokens, which is less than the number of tokens used for pre-training (10B)
> 
> ![Image](https://github.com/user-attachments/assets/acb0b40d-9b31-4597-ac28-7e43f70ad16b)
> 
> However, it is shown in the script that the **continue pretraining** uses the [same training data (from part-000-00000 to part-170-00000)](https://github.com/TsinghuaC3I/Fourier-Position-Embedding/blob/main/configs/c4/length-1024/ce-extra/x%2Byarn/OLMo-60M-ce-yarn.yaml#L283) as [**pre-training**](https://github.com/TsinghuaC3I/Fourier-Position-Embedding/blob/main/configs/c4/length-512/ce-rope/OLMo-60M-ce.yaml#L265)

Oh @hrlics, it seems that the reason may be attributed to that we did not use the full C4. We only utilized 20k steps for `run_pretrain.sh` and 2k steps for `run_finetune.sh`, with the global_train_batch_size is set to 1,024. Or in other words, we employed the checkpoint from the 20k steps as the starting point for the continual pretraining phase.

**Comment by hrlics** [comment]
Hi @Hua-rookie! 

> Oh @hrlics, it seems that the reason may be attributed to that we did not use the full C4. 

Does this mean the full C4 includes `part-000-00000.npy` to `part-170-00000.npy`, but not all of these `npy` files are used during training?


> We only utilized 20k steps for run_pretrain.sh and 2k steps for run_finetune.sh, with the global_train_batch_size is set to 1,024. 

On my end, the `run_pretrain.sh` takes 264k steps and it is the same for `run_finetune.sh`. Could you kindly point out where to set the maximum steps? Because currently, the maximum duration set in [run_pretrain.sh](https://github.com/TsinghuaC3I/Fourier-Position-Embedding/blob/main/configs/c4/length-512/ce-rope/OLMo-60M-ce.yaml#L81) and [run_finetune.sh](https://github.com/TsinghuaC3I/Fourier-Position-Embedding/blob/main/configs/c4/length-1024/ce-extra/x%2Byarn/OLMo-60M-ce-yarn.yaml#L84) are both 1 epoch.

Thanks!

**Comment by Hua-rookie** [comment]
Hi @Hua-rookie,

> Does this mean the full C4 includes part-000-00000.npy to part-170-00000.npy, but not all of these npy files are used during training?

Yes, only part of them are used.

> On my end, the run_pretrain.sh takes 264k steps and it is the same for run_finetune.sh. Could you kindly point out where to set the maximum steps?

We kept the setting unchanged in `run_pretrain.sh` and `run_finetune.sh`, but manually terminated the execution once it reached the 20k steps. This would bring a slower warmup phase, which may be beneficial for pre-training. While this approach may not smart, but it is actually what we did.

**Comment by hrlics** [comment]
> Hi [@Hua-rookie](https://github.com/Hua-rookie),
> 
> > Does this mean the full C4 includes part-000-00000.npy to part-170-00000.npy, but not all of these npy files are used during training?
> 
> Yes, only part of them are used.
> 
> > On my end, the run_pretrain.sh takes 264k steps and it is the same for run_finetune.sh. Could you kindly point out where to set the maximum steps?
> 
> We kept the setting unchanged in `run_pretrain.sh` and `run_finetune.sh`, but manually terminated the execution once it reached the 20k steps. This would bring a slower warmup phase, which may be beneficial for pre-training. While this approach may not smart, but it is actually what we did.

Hi @Hua-rookie 

Thanks a lot for your response! I see the point now. BTW, is it possible to share the scripts for fine-tuning SmolLM-1.7B?

**Comment by Hua-rookie** [comment]
> > Hi [@Hua-rookie](https://github.com/Hua-rookie),
> > > Does this mean the full C4 includes part-000-00000.npy to part-170-00000.npy, but not all of these npy files are used during training?
> > 
> > 
> > Yes, only part of them are used.
> > > On my end, the run_pretrain.sh takes 264k steps and it is the same for run_finetune.sh. Could you kindly point out where to set the maximum steps?
> > 
> > 
> > We kept the setting unchanged in `run_pretrain.sh` and `run_finetune.sh`, but manually terminated the execution once it reached the 20k steps. This would bring a slower warmup phase, which may be beneficial for pre-training. While this approach may not smart, but it is actually what we did.
> 
> Hi [@Hua-rookie](https://github.com/Hua-rookie)
> 
> Thanks a lot for your response! I see the point now. BTW, is it possible to share the scripts for fine-tuning SmolLM-1.7B?

Hi @hrlics,

We fine-tuned SmolLM-1.7B using [huggingface/alignment-handbook](https://github.com/huggingface/alignment-handbook) and data recipes from [huggingface/alignment-handbook/recipes/smollm](https://github.com/huggingface/alignment-handbook/tree/main/recipes/smollm). Prior to fine-tuning, minor modifications should be made to the source code and configuration of SmolLM to accommodate FoPE.

**Comment by hrlics** [comment]
> > > Hi [@Hua-rookie](https://github.com/Hua-rookie),
> > > > Does this mean the full C4 includes part-000-00000.npy to part-170-00000.npy, but not all of these npy files are used during training?
> > > 
> > > 
> > > Yes, only part of them are used.
> > > > On my end, the run_pretrain.sh takes 264k steps and it is the same for run_finetune.sh. Could you kindly point out where to set the maximum steps?
> > > 
> > > 
> > > We kept the setting unchanged in `run_pretrain.sh` and `run_finetune.sh`, but manually terminated the execution once it reached the 20k steps. This would bring a slower warmup phase, which may be beneficial for pre-training. While this approach may not smart, but it is actually what we did.
> > 
> > 
> > Hi [@Hua-rookie](https://github.com/Hua-rookie)
> > Thanks a lot for your response! I see the point now. BTW, is it possible to share the scripts for fine-tuning SmolLM-1.7B?
> 
> Hi [@hrlics](https://github.com/hrlics),
> 
> We fine-tuned SmolLM-1.7B using [huggingface/alignment-handbook](https://github.com/huggingface/alignment-handbook) and data recipes from [huggingface/alignment-handbook/recipes/smollm](https://github.com/huggingface/alignment-handbook/tree/main/recipes/smollm). Prior to fine-tuning, minor modifications should be made to the source code and configuration of SmolLM to accommodate FoPE.

I see. Thanks again!
